File: regression/modeling.R
## block 1
mh <- function(x) {
  xbar <- mean(x)
  crit <- 1 / length(x)^0.25
  if (abs(xbar) > crit) {
    return(xbar)
  } else {
    return(0)
  }
}
## end

## block 2
relvariance <- function(mu, n, nsims = 2000) {
  variances <- apply(replicate(nsims, {
    x <- rnorm(n, mu)
    c(xbar = mean(x), mhat = mh(x))
  }), 1, var)
  return(variances[2] / variances[1])
}
## end

set.seed(647893)
exvar <-
## block 3
relvariance(1, 100)
## end

mcvars <-
## block 4
sapply(seq(-1, 1, 0.01), function(m) relvariance(m, 100))
## end
pdf(file = "regression/modeling_fig1.pdf", width=5, height=5)
plot(mcvars ~ seq(-1, 1, 0.01), type = "l")
lines(c(1,1) ~ c(-1,1), col = "light gray")
dev.off()

mcvars2 <-
## block 5
sapply(seq(-1, 1, 0.01), function(m) relvariance(m, 1000))
## end
pdf(file = "regression/modeling_fig2.pdf", width=5, height=5)
plot(mcvars2 ~ seq(-1, 1, 0.01), type = "l")
lines(c(1,1) ~ c(-1,1), col = "light gray")
dev.off()


texcommand <- function(macro, tex, fmt = "%s")
  sprintf("\\newcommand{\\%s}{%s}\n", macro, sprintf(fmt, tex))

cat(file = "regression/modeling_macros.tex",
    texcommand("exvar", exvar, "%.2f"))

File: regression/modeling.jl
using Gadfly
using Distributions

## block 1
function μh(x)
    xbar = mean(x)
    if abs(xbar) > 1 / length(x)^0.25
        return xbar
    else
        return 0
    end
end
## end

## block 2
function relativevariance(μ, n; nsims = 2000)
    xbar = Array(Real,nsims)
    μhat = Array(Real,nsims)
    for i = 1:nsims
        x = rand(Normal(μ), n)
        xbar[i] = mean(x)
        μhat[i] = μh(x)
    end
    var(μhat) / var(xbar)
end
## end

## block 3
exampleratio = relativevariance(.1, 100)
## end

mcvars =
## block 4
[relativevariance(μ, 100) for μ in -1:0.01:1]
## end

mcvars2 =
## block 5
[relativevariance(μ, 1000) for μ in -1:0.01:1]
## end

function saveplot(id, x, y)
    draw(PDF("analysis/modeling_$id.pdf", 5inch, 3inch),
         plot(x = x, y = y, Geom.line,
              Guide.xlabel("μ"), Guide.ylabel("Relative SD"),
              Theme(default_color = color("black"),
                    panel_fill = color("white"),
                    grid_color = color("white"),
                    minor_label_font = "Bitstream Charter",
                    major_label_font = "Bitstream Charter")))
end

saveplot("fig1", -1:0.01:1, mcvars)
saveplot("fig2", -1:0.01:1, mcvars2)

outfile = open("analysis/modeling_macros.tex", "w")
@printf(outfile, "\\newcommand{\\exampleratio}{%.3f}\n", exampleratio)
close(outfile)
